apisix/plugins/openwhisk.lua

```diff
@@ -0,0 +1,124 @@
+local http              = require("resty.http")
+local ngx_encode_base64 = ngx.encode_base64
+local tostring          = tostring
+
+local schema = {
+    type = "object",
+    properties = {
+        api_host = {type = "string"},
+        ssl_verify = {
+            type = "boolean",
+            default = true,
+        },
+        service_token = {type = "string"},
+        namespace = {type = "string"},
+        action = {type = "string"},
+        result = {
+            type = "boolean",
+            default = true,
+        },
+        timeout = {
+            type = "integer",
+            minimum = 1000,
+            maximum = 60000,
+            default = 3000,
+            description = "timeout in milliseconds",
+        },
+        keepalive = {type = "boolean", default = true},
+        keepalive_timeout = {type = "integer", minimum = 1000, default = 60000},
+        keepalive_pool = {type = "integer", minimum = 1, default = 5}
+    },
+    required = {"api_host", "service_token", "namespace", "action"}
+}
+
+
+local _M = {
+    version = 0.1,
+    priority = -1901,
+    name = "openwhisk",
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    local ok, err = core.schema.check(schema, conf)
+    if not ok then
+        return false, err
+    end
+
+    return true
+end
+
+
+function _M.access(conf, ctx)
+    if core.request.get_method() ~= "GET" then
+        local content_type = core.request.header(ctx, "Content-Type")
+        if not core.string.has_prefix(content_type, "application/json") then
+            core.log.error("only support json request body")
+            return 415
+        end
+    end
+
+    -- Please avoid transferring large amounts of data or uploading files, which can
+    -- cause your requested data to be discarded. Depends on `client_body_buffer_size`.
+    local params = {
+        method = "POST",
+        body = core.request.get_body(),
+        query = {
+            blocking = "true",
+            result = tostring(conf.result),
+            timeout = conf.timeout
+        },
+        headers = {
+            ["Authorization"] = "Basic " .. ngx_encode_base64(conf.service_token),
+            ["Content-Type"] = "application/json",
+        },
+        keepalive = conf.keepalive,
+        ssl_verify = conf.ssl_verify
+    }
+
+    if conf.keepalive then
+        params.keepalive_timeout = conf.keepalive_timeout
+        params.keepalive_pool = conf.keepalive_pool
+    end
+
+    -- OpenWhisk action endpoint
+    local endpoint = conf.api_host .. "/api/v1/namespaces/" .. conf.namespace ..
+        "/actions/" .. conf.action
+
+    local httpc = http.new()
+    httpc:set_timeout(conf.timeout)
+
+    local res, err = httpc:request_uri(endpoint, params)
+
+    if not res or err then
+        core.log.error("failed to process openwhisk action, err: ", err)
+        return 503
+    end
+
+    -- setting response headers
+    core.response.set_header(res.headers)
```

Comment:
Should we handle HTTP2 like the azure version?

Reply:
From current tests, OpenWhisk does not generate these headers.